### PR TITLE
59 randomized svd

### PR DESCRIPTION
**Pull request by bef-18**
Randomized SVD is implemented. POD can be done now through randomized SVD. The possibility of projecting the data for DMD with a randomized SVD will be added through the optimized DMD branch, where a major refracturing of DMD will be done.

### CODE REVIEW

File: pyLOM/vmmath/src/svd.c [hunk 1]
```diff
@@ -1009,6 +1009,142 @@ int ztsqr_svd(dcomplex_t *Ui, double *S, dcomplex_t *VT, dcomplex_t *Ai, const i
 	return info;
 }
 
+int srandomized_svd(float *Ui, float *S, float *VT, float *Ai, const int m, const int n, const int r, const int q, MPI_Comm comm) {
+	/*
+		Randomized single value decomposition (SVD) with oversampling and power iterations with the algorithm from
```
Comment (ArnauMiro) [inline pyLOM/vmmath/src/svd.c:1014]: Missing proper reference?

File: pyLOM/vmmath/src/svd.c [hunk 2]
```diff
@@ -1009,6 +1009,142 @@ int ztsqr_svd(dcomplex_t *Ui, double *S, dcomplex_t *VT, dcomplex_t *Ai, const i
...
+	
+	free(Ur); free(Qi); 
+
+	return info;
+}
+
+int drandomized_svd(double *Ui, double *S, double *VT, double *Ai, const int m, const int n, const int r, const int q, MPI_Comm comm) {
+	/*
+		Randomized single value decomposition (SVD) with oversampling and power iterations with the algorithm from
+
```
Comment (ArnauMiro) [inline pyLOM/vmmath/src/svd.c:1083]: same as before

File: pyLOM/vmmath/src/vector_matrix.c [hunk 3]
```diff
@@ -419,17 +421,17 @@ int ceigen(float *real, float *imag, scomplex_t *w, float *A,
 	vecs = (float*)malloc(n*n*sizeof(float));
 	info = LAPACKE_sgeev(
 		LAPACK_ROW_MAJOR, // int  		matrix_layout
-		           'N',   // char       jobvl
```
Comment (ArnauMiro) [inline pyLOM/vmmath/src/vector_matrix.c:422]: Alineation being undone, please fix

File: pyLOM/vmmath/src/vector_matrix.c [hunk 4]
```diff
@@ -473,17 +475,17 @@ int zeigen(double *real, double *imag, dcomplex_t *w, double *A,
 	vecs = (double*)malloc(n*n*sizeof(double));
 	info = LAPACKE_dgeev(
 		LAPACK_ROW_MAJOR, // int  		matrix_layout
-		           'N',   // char       jobvl
```
Comment (ArnauMiro) [inline pyLOM/vmmath/src/vector_matrix.c:476]: Alineation being undone, please fix

File: pyLOM/vmmath/src/vector_matrix.c [hunk 5]
```diff
@@ -573,10 +575,10 @@ int ccholesky(scomplex_t *A, int N){
 	int info, ii, jj;
 	info = LAPACKE_cpotrf(
 		LAPACK_ROW_MAJOR, // int  	matrix_layout
-		             'L', //char	Decide if the Upper or the Lower triangle of A are stored
```
Comment (ArnauMiro) [inline pyLOM/vmmath/src/vector_matrix.c:576]: Alineation being undone, please fix

File: pyLOM/vmmath/src/vector_matrix.c [hunk 6]
```diff
@@ -595,10 +597,10 @@ int zcholesky(dcomplex_t *A, int N){
 	int info, ii, jj;
 	info = LAPACKE_zpotrf(
 		LAPACK_ROW_MAJOR, // int  	matrix_layout
-		             'L', //char	Decide if the Upper or the Lower triangle of A are stored
```
Comment (ArnauMiro) [inline pyLOM/vmmath/src/vector_matrix.c:598]: Alineation being undone, please fix

File: pyLOM/vmmath/src/vector_matrix.c [hunk 7]
```diff
@@ -677,11 +679,11 @@ int sinverse(float *A, int N, char *UoL){
 	int info;
 	info = LAPACKE_strtri(
 		LAPACK_ROW_MAJOR, //int     matrix_layout
-		            *UoL, //char    Decide if the Upper or the Lower triangle of A are stored
```
Comment (ArnauMiro) [inline pyLOM/vmmath/src/vector_matrix.c:680]: Alineation being undone, please fix

File: pyLOM/vmmath/src/vector_matrix.c [hunk 8]
```diff
@@ -246,19 +248,19 @@ void zmatmult(dcomplex_t *C, dcomplex_t *A, dcomplex_t *B, const int m, const in
 	if (*TB == 'C'){ TransB = CblasConjTrans; ldb = k; }
 	cblas_zgemm(
 		CblasRowMajor, // const CBLAS_LAYOUT 	  layout
-		       TransA, // const CBLAS_TRANSPOSE   TransA
```
Comment (ArnauMiro) [inline pyLOM/vmmath/src/vector_matrix.c:249]: Alineation being undone, please fix

File: pyLOM/vmmath/src/vector_matrix.c [hunk 9]
```diff
@@ -213,19 +215,19 @@ void cmatmult(scomplex_t *C, scomplex_t *A, scomplex_t *B, const int m, const in
 	if (*TB == 'C'){ TransB = CblasConjTrans; ldb = k; }
 	cblas_cgemm(
 		CblasRowMajor, // const CBLAS_LAYOUT 	  layout
-		       TransA, // const CBLAS_TRANSPOSE   TransA
```
Comment (ArnauMiro) [inline pyLOM/vmmath/src/vector_matrix.c:216]: Alineation being undone, please fix

File: pyLOM/vmmath/src/vector_matrix.c [hunk 10]
```diff
@@ -160,19 +162,19 @@ void dmatmult(double *C, double *A, double *B, const int m, const int n, const i
 	if (*TB == 'T') {TransB = CblasTrans; ldb = k;}
 	cblas_dgemm(
 		CblasRowMajor, // const CBLAS_LAYOUT 	  layout
-		       TransA, // const CBLAS_TRANSPOSE   TransA
```
Comment (ArnauMiro) [inline pyLOM/vmmath/src/vector_matrix.c:163]: Alineation being undone, please fix

File: pyLOM/vmmath/src/vector_matrix.c [hunk 11]
```diff
@@ -127,19 +129,19 @@ void smatmult(float *C, float *A, float *B, const int m, const int n, const int
 	if (*TB == 'T') {TransB = CblasTrans; ldb = k;}
 	cblas_sgemm(
 		CblasRowMajor, // const CBLAS_LAYOUT 	  layout
-		       TransA, // const CBLAS_TRANSPOSE   TransA
```
Comment (ArnauMiro) [inline pyLOM/vmmath/src/vector_matrix.c:130]: Alineation being undone, please fix

File: pyLOM/vmmath/src/vector_matrix.c [hunk 12]
```diff
@@ -693,11 +695,11 @@ int dinverse(double *A, int N, char *UoL){
 	int info;
 	info = LAPACKE_dtrtri(
 		LAPACK_ROW_MAJOR, //int     matrix_layout
-		            *UoL, //char    Decide if the Upper or the Lower triangle of A are stored
```
Comment (ArnauMiro) [inline pyLOM/vmmath/src/vector_matrix.c:696]: Alineation being undone, please fix

File: pyLOM/vmmath/src/vector_matrix.c [hunk 13]
```diff
@@ -709,11 +711,11 @@ int cinverse(scomplex_t *A, int N, char *UoL){
 	int info;
 	info = LAPACKE_ctrtri(
 		LAPACK_ROW_MAJOR, //int     matrix_layout
-		            *UoL, //char    Decide if the Upper or the Lower triangle of A are stored
```
Comment (ArnauMiro) [inline pyLOM/vmmath/src/vector_matrix.c:712]: Alineation being undone, please fix

File: pyLOM/vmmath/src/vector_matrix.c [hunk 14]
```diff
@@ -725,11 +727,11 @@ int zinverse(dcomplex_t *A, int N, char *UoL){
 	int info;
 	info = LAPACKE_ztrtri(
 		LAPACK_ROW_MAJOR, //int     matrix_layout
-		            *UoL, //char    Decide if the Upper or the Lower triangle of A are stored
```
Comment (ArnauMiro) [inline pyLOM/vmmath/src/vector_matrix.c:728]: Alineation being undone, please fix

File: pyLOM/vmmath/src/vector_matrix.c [hunk 15]
```diff
@@ -849,3 +851,31 @@ void zsort(dcomplex_t *v, int *index, int n){
 	}
 	free(w);
 }
+
+void srandom_matrix(float *A, int m, int n){
```
Comment (ArnauMiro) [inline pyLOM/vmmath/src/vector_matrix.c:855]: Enter seed as an input

File: pyLOM/vmmath/src/vector_matrix.c [hunk 16]
```diff
@@ -849,3 +851,31 @@ void zsort(dcomplex_t *v, int *index, int n){
...
+	srand((unsigned int)time(NULL));
+
+	for (int i = 0; i < m; i++) {
+		for (int j = 0; j < n; j++) {
+			AC_MAT(A,n,i,j) = (float)(rand() / RAND_MAX);
+		}
+	}
+}
+
+void drandom_matrix(double *A, int m, int n){
```
Comment (ArnauMiro) [inline pyLOM/vmmath/src/vector_matrix.c:869]: enter seed as an input

File: pyLOM/vmmath/src/vector_matrix.c [hunk 17]
```diff
@@ -5,6 +5,8 @@
 #include <complex.h>
 #include <stdio.h>
 #include <string.h>
+#include <stdlib.h>
+#include <time.h>
```
Comment (ArnauMiro) [inline pyLOM/vmmath/src/vector_matrix.c:9]: time is not needed since the seed should come as an input.

**Review by ArnauMiro — CHANGES_REQUESTED** [review]

**Comment by bef-18** [comment]
Changes pushed

File: pyLOM/POD/wrapper.pyx [hunk 18]
```diff
@@ -39,16 +39,18 @@ cdef extern from "averaging.h":
 	cdef void c_dsubtract_mean "dsubtract_mean"(double *out, double *X, double *X_mean, const int m, const int n)
```
Comment (ArnauMiro) [inline pyLOM/POD/wrapper.pyx:39]: Alignment issues again?

File: pyLOM/vmmath/src/vector_matrix.c (hunk 11 above)
Comment (ArnauMiro) [reply pyLOM/vmmath/src/vector_matrix.c:130]: Not fixed

File: pyLOM/vmmath/src/vector_matrix.c (hunk 10 above)
Comment (ArnauMiro) [reply pyLOM/vmmath/src/vector_matrix.c:163]: Not fixed

File: pyLOM/vmmath/src/vector_matrix.c (hunk 8 above)
Comment (ArnauMiro) [reply pyLOM/vmmath/src/vector_matrix.c:249]: not fixed

File: pyLOM/vmmath/src/vector_matrix.c (hunk 3 above)
Comment (ArnauMiro) [reply pyLOM/vmmath/src/vector_matrix.c:422]: Not fixed

File: pyLOM/vmmath/src/vector_matrix.c (hunk 4 above)
Comment (ArnauMiro) [reply pyLOM/vmmath/src/vector_matrix.c:476]: Not fixed

File: pyLOM/vmmath/src/vector_matrix.c (hunk 5 above)
Comment (ArnauMiro) [reply pyLOM/vmmath/src/vector_matrix.c:576]: not fixed

File: pyLOM/vmmath/src/vector_matrix.c (hunk 6 above)
Comment (ArnauMiro) [reply pyLOM/vmmath/src/vector_matrix.c:598]: Not fixed

File: pyLOM/vmmath/src/vector_matrix.c (hunk 7 above)
Comment (ArnauMiro) [reply pyLOM/vmmath/src/vector_matrix.c:680]: Not fixed

File: pyLOM/vmmath/src/vector_matrix.c (hunk 12 above)
Comment (ArnauMiro) [reply pyLOM/vmmath/src/vector_matrix.c:696]: Not fixed

File: pyLOM/vmmath/src/vector_matrix.c (hunk 13 above)
Comment (ArnauMiro) [reply pyLOM/vmmath/src/vector_matrix.c:712]: Not fixed

File: pyLOM/vmmath/src/vector_matrix.c (hunk 14 above)
Comment (ArnauMiro) [reply pyLOM/vmmath/src/vector_matrix.c:728]: Not fixed

File: pyLOM/vmmath/src/svd.c [hunk 19]
```diff
@@ -1009,6 +1010,148 @@ int ztsqr_svd(dcomplex_t *Ui, double *S, dcomplex_t *VT, dcomplex_t *Ai, const i
...
+		Ui(m,n)  POD modes dispersed on each processor (must come preallocated).
+		S(n)     singular values.
+		VT(n,n)  right singular vectors (transposed).
+	*/
+	int info = 0;
+	int ii   = 0;
+	// Multiply per a random matrix
+	float *omega;
+	float *Y;
+	unsigned int seed = (unsigned int)time(NULL);
```
Comment (ArnauMiro) [inline pyLOM/vmmath/src/svd.c:1030]: I'd keep the seed outside and would set this up at the cython layer, to keep the C layer clean.

**Review by ArnauMiro — CHANGES_REQUESTED** [review]

File: pyLOM/vmmath/src/svd.c (hunk 19 above)
Comment (bef-18) [reply pyLOM/vmmath/src/svd.c:1030]: Shall I put always the same number for reproducibility in the cython layer? or a random int each time?

File: pyLOM/POD/wrapper.pyx [hunk 20]
```diff
@@ -79,11 +81,13 @@ def _srun(float[:,:] X, int remove_mean):
 	cdef int m = X.shape[0], n = X.shape[1], mn = min(m,n), retval
 	cdef float *X_mean
 	cdef float *Y
+	cdef unsigned int seed = 19
```
Comment (ArnauMiro) [inline pyLOM/POD/wrapper.pyx:84]: Well here is where you should call the time module...
```
from libc.time   cimport time
cdef unsigned int seed = <int>time(NULL)
```

File: pyLOM/POD/wrapper.pyx [hunk 21]
```diff
@@ -127,11 +134,13 @@ def _drun(double[:,:] X, int remove_mean):
 	cdef int m = X.shape[0], n = X.shape[1], mn = min(m,n), retval
 	cdef double *X_mean
 	cdef double *Y
+	cdef unsigned int seed = 19
```
Comment (ArnauMiro) [inline pyLOM/POD/wrapper.pyx:137]: same as before

File: pyLOM/vmmath/src/vector_matrix.c
Comment (ArnauMiro) [inline pyLOM/vmmath/src/vector_matrix.c:1]: @bef-18 recheck the indentations of this whole file here in github, they look weird and not solved

File: pyLOM/vmmath/wrapper.pyx [hunk 22]
```diff
@@ -1106,6 +1108,69 @@ def tsqr_svd(real_full[:,:] A):
...
+def _srandomized_svd(float[:,:] A, int r, int q):
+	'''
+	Parallel Randomized Single value decomposition (SVD) using Lapack.
+		U(m,n)   are the POD modes.
+		S(n)     are the singular values.
+		V(n,n)   are the right singular vectors.
+	'''
+	cdef int retval
+	cdef int m = A.shape[0], n = A.shape[1], mn = min(m,n)
+	cdef unsigned int seed = 19
```
Comment (ArnauMiro) [inline pyLOM/vmmath/wrapper.pyx:1124]: seed please

File: pyLOM/vmmath/wrapper.pyx [hunk 23]
```diff
@@ -1106,6 +1108,69 @@ def tsqr_svd(real_full[:,:] A):
...
+def _drandomized_svd(double[:,:] A, int r, int q):
+	'''
+	Parallel Randomized Single value decomposition (SVD) using Lapack.
+		U(m,n)   are the POD modes.
+		S(n)     are the singular values.
+		V(n,n)   are the right singular vectors.
+	'''
+	cdef int retval
+	cdef int m = A.shape[0], n = A.shape[1], mn = min(m,n)
+	cdef unsigned int seed = 19
```
Comment (ArnauMiro) [inline pyLOM/vmmath/wrapper.pyx:1147]: seed please

**Review by ArnauMiro — CHANGES_REQUESTED** [review]

File: pyLOM/POD/wrapper.pyx (hunk 20 above)
Comment (bef-18) [reply pyLOM/POD/wrapper.pyx:84]: I just wanted the seed to be Lamine Yamal's number :(

File: pyLOM/vmmath/wrapper.pyx (hunk 22 above)
Comment (ArnauMiro) [reply pyLOM/vmmath/wrapper.pyx:1124]: We cannot have a fixed seed for random number generation

File: pyLOM/vmmath/wrapper.pyx (hunk 23 above)
Comment (ArnauMiro) [reply pyLOM/vmmath/wrapper.pyx:1147]: Same

File: pyLOM/POD/wrapper.pyx (hunk 21 above)
Comment (ArnauMiro) [reply pyLOM/POD/wrapper.pyx:137]: Not fixed

File: pyLOM/POD/wrapper.pyx (hunk 20 above)
Comment (ArnauMiro) [reply pyLOM/POD/wrapper.pyx:84]: Please fix

**Review by ArnauMiro — CHANGES_REQUESTED** [review]

**Review by ArnauMiro — APPROVED** [review]

**Review by ArnauMiro — DISMISSED** [review]

**Review by ArnauMiro — APPROVED** [review]

**Pull request merged**